data/crac/crac-impl/src/main/java/com/powsybl/openrao/data/cracimpl/NetworkActionImpl.java

```diff
@@ -7,55 +7,116 @@
+    @Override
+    public Set<Action> getElementaryActions() {
         return elementaryActions;
     }
 
     @Override
     public boolean hasImpactOnNetwork(Network network) {
-        return elementaryActions.stream().anyMatch(elementaryAction -> elementaryAction.hasImpactOnNetwork(network));
+        return elementaryActions.stream().anyMatch(elementaryAction -> {
+            if (elementaryAction instanceof GeneratorAction generatorAction) {
+                Generator generator = network.getGenerator(generatorAction.getGeneratorId());
+                return Math.abs(generator.getTargetP() - generatorAction.getActivePowerValue().getAsDouble()) >= EPSILON;
+            } else if (elementaryAction instanceof LoadAction loadAction) {
+                Load load = network.getLoad(loadAction.getLoadId());
+                return Math.abs(load.getP0() - loadAction.getActivePowerValue().getAsDouble()) >= EPSILON;
+            } else if (elementaryAction instanceof DanglingLineAction danglingLineAction) {
+                DanglingLine danglingLine = network.getDanglingLine(danglingLineAction.getDanglingLineId());
+                return Math.abs(danglingLine.getP0() - danglingLineAction.getActivePowerValue().getAsDouble()) >= EPSILON;
+            } else if (elementaryAction instanceof ShuntCompensatorPositionAction shuntCompensatorPositionAction) {
+                ShuntCompensator shuntCompensator = network.getShuntCompensator(shuntCompensatorPositionAction.getShuntCompensatorId());
+                return Math.abs(shuntCompensator.getSectionCount() - shuntCompensatorPositionAction.getSectionCount()) > 0;
+            } else if (elementaryAction instanceof PhaseTapChangerTapPositionAction phaseTapChangerTapPositionAction) {
+                PhaseTapChanger phaseTapChanger = network.getTwoWindingsTransformer(phaseTapChangerTapPositionAction.getTransformerId()).getPhaseTapChanger();
+                return phaseTapChangerTapPositionAction.getTapPosition() != phaseTapChanger.getTapPosition();
+            } else if (elementaryAction instanceof SwitchPair switchPair) {
+                return !network.getSwitch(switchPair.getSwitchToOpen().getId()).isOpen() || network.getSwitch(switchPair.getSwitchToClose().getId()).isOpen();
+            } else if (elementaryAction instanceof TerminalsConnectionAction terminalsConnectionAction) {
+                Connectable<?> connectable = network.getConnectable(terminalsConnectionAction.getElementId());
+                if (terminalsConnectionAction.isOpen()) {
+                    // Connectable is considered closed if all terminals are connected
+                    return connectable.getTerminals().stream().allMatch(Terminal::isConnected);
+                } else {
+                    // Connectable is already considered opened if one of the terminals is disconnected
+                    return !connectable.getTerminals().stream().allMatch(Terminal::isConnected);
```

Comment:
To be checked by @pet-mit

Reply:
As for other use case of Connectable, I should put back the original Branch code instead:

```
        Identifiable<?> element = network.getIdentifiable(networkElement.getId());
        if (element instanceof Branch<?> branch) {
            if (actionType == ActionType.OPEN) {
                // Line is considered closed if both terminal are connected
                return branch.getTerminal1().isConnected() && branch.getTerminal2().isConnected();
            } else {
                // Line is already considered opened if one of the terminals is disconnected
                return !branch.getTerminal1().isConnected() || !branch.getTerminal2().isConnected();
            }
        } else if (element instanceof Switch sw) {
            return sw.isOpen() == (actionType == ActionType.CLOSE);
        } else {
            throw new NotImplementedException("Topological actions are only on branches or switches for now");
        }
```

Reply:
fixed Connectable -> Branch